File: src/main/scala/com/dyonovan/neotech/api/jei/alloyer/JEIAlloyerRecipeCategory.scala
package com.dyonovan.neotech.api.jei.alloyer

import java.awt.Color

import com.dyonovan.neotech.api.jei.NeoTechPlugin
import com.dyonovan.neotech.lib.Reference
import com.teambr.bookshelf.api.jei.drawables.{GuiComponentArrowJEI, GuiComponentBox, GuiComponentPowerBarJEI}
import mezz.jei.api.gui.{IDrawable, IRecipeLayout}
import mezz.jei.api.recipe.{IRecipeCategory, IRecipeWrapper}
import net.minecraft.client.Minecraft
import net.minecraft.util.ResourceLocation
import net.minecraft.util.text.translation.I18n

/**
  * This file was created for NeoTech
  *
  * NeoTech is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Dyonovan
  * @since 2/22/2016
  */
class JEIAlloyerRecipeCategory extends IRecipeCategory {

    val location = new ResourceLocation(Reference.MOD_ID, "textures/gui/jei/jei.png")
    val arrow = new GuiComponentArrowJEI(81, 17, NeoTechPlugin.jeiHelpers)
    val power = new GuiComponentPowerBarJEI(14, 0, 18, 60, new Color(255, 0, 0), NeoTechPlugin.jeiHelpers) {
        addColor(new Color(255, 150, 0))
        addColor(new Color(255, 255, 0))
    }
    val tankIn1 = new GuiComponentBox(38, 0, 18, 60)
    val tankIn2 = new GuiComponentBox(60, 0, 18, 60)
    val tankOut = new GuiComponentBox(115, 0, 50, 60)

    override def getBackground: IDrawable = NeoTechPlugin.jeiHelpers.getGuiHelper.createDrawable(location, 0, 0, 170, 60)

    override def setRecipe(recipeLayout: IRecipeLayout, recipeWrapper: IRecipeWrapper): Unit = {
        val fluids = recipeLayout.getFluidStacks
        fluids.init(0, true, 39, 0, 16, 59, 2000, false, null)
        fluids.init(1, true, 61, 0, 16, 59, 2000, false, null)
        fluids.init(2, false, 116, 0, 48, 59, 2000, false, null)

        recipeWrapper match {
            case alloyer: JEIAlloyerRecipe =>
                recipeLayout.getFluidStacks.set(0, alloyer.getFluidInputs.get(0))
                recipeLayout.getFluidStacks.set(1, alloyer.getFluidInputs.get(1))
                recipeLayout.getFluidStacks.set(2, alloyer.getFluidOutputs)
        }
    }

    override def drawAnimations(minecraft: Minecraft): Unit = {
        arrow.draw(minecraft, 0, 0)
        power.draw(minecraft, 0, 0)
    }

    override def drawExtras(minecraft: Minecraft): Unit = {
        tankIn1.draw(minecraft)
        tankIn2.draw(minecraft)
        tankOut.draw(minecraft)
    }

    override def getTitle: String = I18n.translateToLocal("tile.neotech:alloyer.name")

    override def getUid: String = Reference.MOD_ID + ":alloyer"
}

File: src/main/java/com/teambrmodding/neotech/api/jei/alloyer/JEIAlloyerRecipeCategory.java
package com.teambrmodding.neotech.api.jei.alloyer;

import com.teambr.bookshelf.helper.LogHelper;
import com.teambr.bookshelf.util.ClientUtils;
import com.teambrmodding.neotech.api.jei.NeotechJEIPlugin;
import com.teambrmodding.neotech.lib.Reference;
import com.teambrmodding.neotech.managers.RecipeManager;
import com.teambrmodding.neotech.registries.AbstractRecipe;
import com.teambrmodding.neotech.registries.AlloyerRecipeHandler;
import mezz.jei.api.gui.*;
import mezz.jei.api.ingredients.IIngredients;
import mezz.jei.api.recipe.IRecipeCategory;
import net.minecraft.client.Minecraft;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.fluids.FluidStack;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.List;

/**
 * This file was created for NeoTech
 *
 * NeoTech is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis - pauljoda
 * @since 2/5/2017
 */
public class JEIAlloyerRecipeCategory implements IRecipeCategory<JEIAlloyerRecipeWrapper> {

    // Display
    private ResourceLocation backgroundResource = new ResourceLocation(Reference.MOD_ID, "textures/gui/jei/electricAlloyer.png");
    private IDrawableAnimated progressArrow;
    private IDrawableAnimated powerBar;

    /*******************************************************************************************************************
     * Constructor                                                                                                     *
     *******************************************************************************************************************/

    /**
     * Constructor
     */

    public JEIAlloyerRecipeCategory() {
        IDrawableStatic progressArrowDrawable = NeotechJEIPlugin.jeiHelpers.getGuiHelper().createDrawable(backgroundResource, 169, 0, 23, 17);
        progressArrow = NeotechJEIPlugin.jeiHelpers.getGuiHelper().createAnimatedDrawable(progressArrowDrawable, 200, IDrawableAnimated.StartDirection.LEFT, false);

        IDrawableStatic powerBarDrawable = NeotechJEIPlugin.jeiHelpers.getGuiHelper().createDrawable(backgroundResource, 169, 17, 16, 62);
        powerBar = NeotechJEIPlugin.jeiHelpers.getGuiHelper().createAnimatedDrawable(powerBarDrawable, 300, IDrawableAnimated.StartDirection.TOP, true);
    }

    /*******************************************************************************************************************
     * IRecipeCategory                                                                                                 *
     *******************************************************************************************************************/

    /**
     * Get the unique ID of this category
     * @return The alloyer string
     */
    @Override
    public String getUid() {
        return NeotechJEIPlugin.ALLOYER_UUID;
    }

    /**
     * Used to display the title
     * @return The translated title
     */

    @Override
    public String getTitle() {
        return ClientUtils.translate("tile.neotech:alloyer.name");
    }

    /**
     * Set background, we just use a flat gray, might expand in future to a texture
     * @return A drawable made from our texture
     */
    @Override
    public IDrawable getBackground() {
        return NeotechJEIPlugin.jeiHelpers.getGuiHelper().createDrawable(backgroundResource, 0, 0, 169, 80);
    }

    /**
     * We will use the default one registered in the NeotechJEIPlugin class
     * @return Null, don't worry about it here
     */
    @Nullable
    @Override
    public IDrawable getIcon() {
        return null;
    }

    /**
     * The main draw call, display generic stuff here
     */
    @Override
    public void drawExtras(Minecraft minecraft) {
        // Draw Animations
        progressArrow.draw(minecraft, 78, 32);
        powerBar.draw(minecraft, 12, 9);
    }

    @Override
    public void drawAnimations(Minecraft minecraft) {
        // Deprecated, moved to drawExtras
    }

    @Override
    public void setRecipe(IRecipeLayout recipeLayout, JEIAlloyerRecipeWrapper recipeWrapper) {
        // Deprecated
    }

    /**
     * Actually exposes the recipe to JEI
     * @param recipeLayout The object to hold al the info
     * @param recipeWrapper The recipe, not always needed as all ingredients are added to ingredients
     * @param ingredients What holds the ingredients
     */
    @Override
    public void setRecipe(IRecipeLayout recipeLayout, JEIAlloyerRecipeWrapper recipeWrapper, IIngredients ingredients) {
        IGuiFluidStackGroup fluidStackGroup = recipeLayout.getFluidStacks();

        // Load the fluids
        fluidStackGroup.init(0, true,  36, 9, 16, 62, 2000, false, null);
        fluidStackGroup.init(1, true,  57, 9, 16, 62, 2000, false, null);
        fluidStackGroup.init(2, false, 108, 9, 49, 62, 2000, false, null);

        // Set into layout
        recipeLayout.getFluidStacks().set(0, ingredients.getInputs(FluidStack.class).get(0));
        recipeLayout.getFluidStacks().set(1, ingredients.getInputs(FluidStack.class).get(1));
        recipeLayout.getFluidStacks().set(2, ingredients.getOutputs(FluidStack.class).get(0));
    }


    /*******************************************************************************************************************
     * Class Methods                                                                                                   *
     *******************************************************************************************************************/

    /**
     * Used to generate a list of all recipes for this category
     * @return The completed list of recipes
     */
    public static List<JEIAlloyerRecipeWrapper> buildRecipeList() {
        ArrayList<JEIAlloyerRecipeWrapper> recipes = new ArrayList<>();
        AlloyerRecipeHandler alloyerRecipeHandler = RecipeManager.getHandler(RecipeManager.RecipeType.ALLOYER);
        for(AlloyerRecipeHandler.AlloyerRecipe recipe : alloyerRecipeHandler.recipes) {
            FluidStack fluidInputOne = AbstractRecipe.getFluidStackFromString(recipe.fluidStackOne);
            FluidStack fluidInputTwo = AbstractRecipe.getFluidStackFromString(recipe.fluidStackTwo);
            FluidStack fluidOutput   = AbstractRecipe.getFluidStackFromString(recipe.fluidStackOutput);
            if(fluidInputOne != null && fluidInputTwo != null && fluidOutput != null)
                recipes.add(new JEIAlloyerRecipeWrapper(fluidInputOne, fluidInputTwo, fluidOutput));
            else
                LogHelper.logger.error("[Neotech] Alloyer Recipe json is corrupt! Please delete and run again.");
        }

        return recipes;
    }
}
